fl4health/parameter_exchange/sparse_coo_parameter_exchanger.py

```diff
@@ -12,117 +12,114 @@
-        self, model: nn.Module, initial_model: Optional[nn.Module] = None
-    ) -> Tuple[NDArrays, Tuple[NDArrays, NDArrays, List[str]]]:
-        """
-        Generate masks for model's parameters and select those with mask values 1
-        for exchange. Also extract all the information required to represent
+        Finally, the method extracts all the information required to represent
         the selected parameters in the sparse COO tensor format. More specifically,
         the information consists of the indices of the parameters within the tensor
         to which they belong, the shape of that tensor, and also the name of it.
 
         Args:
             model (nn.Module): Current model.
-            initial_model (Optional[nn.Module], optional): Initial model. Defaults to None.
+            initial_model (nn.Module): Initial model.
 
         Returns:
             Tuple[NDArrays, Tuple[NDArrays, NDArrays, List[str]]]: the selected parameters
             and other information, as detailed above.
         """
-        tensor_names_to_masks = self.create_masks(model, initial_model)
-        model_states = model.state_dict()
+        all_parameter_scores = self.generate_parameter_scores(model, initial_model)
+        all_scores = torch.cat([val.flatten() for _, val in all_parameter_scores.items()])
+        # Sorting all scores and determining the threshold.
+        sorted_scores, _ = torch.sort(all_scores, descending=True)
+        n_top_scores = math.ceil(len(sorted_scores) * self.sparsity_level)
+        # Sanity check.
+        assert n_top_scores >= 1
+        top_scores = sorted_scores[:n_top_scores]
+        score_threshold = top_scores[-1].item()
 
+        # Apply the score threshold to each model tensor to obtain the corresponding sparse tensor.
         selected_parameters_all_tensors = []
         selected_indices_all_tensors = []
         tensor_shapes = []
-        tensor_names = list(tensor_names_to_masks.keys())
-
-        # Note that param_names_to_masks only contains those tensors whose
-        # mask values are not all zero.
-        for tensor_name, tensor_mask in tensor_names_to_masks.items():
+        tensor_names = []
+        model_states = model.state_dict()
+        for tensor_name, param_scores in all_parameter_scores.items():
             model_tensor = model_states[tensor_name]
-            assert model_tensor.shape == tensor_mask.shape
-            sparse_tensor = torch.multiply(model_tensor, tensor_mask)
-            selected_parameters, selected_indices, tensor_shape = self.parameter_packer.extract_coo_info_from_dense(
-                sparse_tensor
-            )
+            # Sanity check.
+            assert model_tensor.shape == param_scores.shape
 
-            selected_parameters_all_tensors.append(selected_parameters)
-            selected_indices_all_tensors.append(selected_indices)
-            tensor_shapes.append(tensor_shape)
+            unique_score_values = torch.unique(
```

Comment:
I'd suggest pulling this check/warning into a helper function so as not to clutter the calculations here